### ISSUE TITLE
ZIM files larger than 2GB can not be handled by the wasm version of libzim/kiwix-lib

### ISSUE THREAD

**Issue by mossroy** [kiwix/kiwix-js#513]
(description updated on 2021-10-30)

It's a serious problem as there are a lot of ZIM files larger than 2GB. For example : https://download.kiwix.org/zim/wikipedia/?C=S;O=D

A first upstream bug was filed : https://github.com/emscripten-core/emscripten/issues/5250 and is marked as fixed.
But it did not seem to be enough for our need.
I opened another one : https://github.com/emscripten-core/emscripten/issues/15394

**Comment by Jaifroid** [comment]
Ah good, I read that thread with interest, as I was worried that maybe the whole file was being read into memory. Good to hear it may be an arbitrary limitation, and that WORKERFS doesn't work by reading the whole file into memory.

**Comment by mossroy** [comment]
We should test again with a recent version of emscripten (>=1.38.35) because https://github.com/emscripten-core/emscripten/pull/8759 might have solved that

**Comment by mossroy** [comment]
I'll need to switch to another emscripten docker image : https://hub.docker.com/r/trzeci/emscripten
But the recent emscripten versions are not available, they're working on it : https://github.com/trzecieu/emscripten-docker/issues/46

**Comment by mossroy** [comment]
I created a new branch `emscripten-upgrade-for-large-file-support` (https://github.com/mossroy/libzim_wasm/tree/emscripten-upgrade-for-large-file-support) using this other emscripten docker image (version 1.38.43)
Our docker image can be built and used, but emscripten compilation of our code fails
with :
```
warning: unexpected argument type i32 at index 1 in call to '__fseeko',
 should be i64
warning: unexpected return type i32 in call to '__ftello', should be i6
4
warning: unexpected argument type i32 at index 1 in call to '__fseeko',
 should be i64
warning: unexpected argument type i32 at index 1 in call to '__fseeko',
 should be i64
warning: unexpected argument type i32 at index 3 in call to 'pread', sh
ould be i64
warning: invalid item (maybe a typo?) in EXPORTED_RUNTIME_METHODS: ALLOC_STATIC
error: asm2wasm seeing an invalid argument type at index 1 (this will not valida
te) in call from $__ZNSt3__213basic_filebufIcNS_11char_traitsIcEEE7seekoffExNS_8
ios_base7seekdirEj to $___fseeko (this is likely due to undefined behavior in C,
 like defining a function one way and calling it in another, which is important
to fix)
error: asm2wasm seeing an invalid argument type at index 1 (this will not valida
te) in call from $__ZNSt3__213basic_filebufIcNS_11char_traitsIcEEE7seekposENS_4f
posI11__mbstate_tEEj to $___fseeko (this is likely due to undefined behavior in 
C, like defining a function one way and calling it in another, which is importan
t to fix)
error: asm2wasm seeing an invalid argument type at index 1 (this will not valida
te) in call from $__ZNSt3__213basic_filebufIcNS_11char_traitsIcEEE4syncEv to $__
_fseeko (this is likely due to undefined behavior in C, like defining a function
 one way and calling it in another, which is important to fix)
error: asm2wasm seeing an invalid argument type at index 3 (this will not valida
te) in call from $__ZNK3zim4unix2FD6readAtEPcNS_7zsize_tENS_8offset_tE to $_prea
d (this is likely due to undefined behavior in C, like defining a function one w
ay and calling it in another, which is important to fix)
[wasm-validator error in function $__ZNSt3__213basic_filebufIcNS_11char_traitsIc
EEE7seekoffExNS_8ios_base7seekdirEj] 1 != 2: call param types must match, on 
[i32] (call $___fseeko
 [i32] (local.get $52)
 [i32] (local.get $60)
 [i32] (local.get $62)
)
...
```

**Comment by mossroy** [comment]
The error message above was certainly because I did not `make clean` before `make` : shame on me.
But it now fails on compiling curl : 

    checking for sys/socket.h... (cached) yes
    checking for recv... no
    configure: error: Unable to link function recv
    Makefile:65: recipe for target 'curlbuild' failed
    make: *** [curlbuild] Error 1

**Comment by mossroy** [comment]
It seems to depend on the version of emscripten.
I do not have this error with version 1.38.35 : curl can be compiled, but I have another issue later :

    em++ --bind demo_file_api.cpp libzimbuild/lib/libzim.so kiwixlibbuild/lib/libkiwix.so -Ilibzimbuild/include -Ikiwixlibbuild/include -Iicubuild/include -fdiagnostics-color=always -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -O0 -g -D_LARGEFILE64_SOURCE=1 -D_FILE_OFFSET_BITS=64 -pthread --pre-js prejs_file_api.js --post-js postjs_file_api.js -s DISABLE_EXCEPTION_CATCHING=0 -s "EXTRA_EXPORTED_RUNTIME_METHODS=['ALLOC_NORMAL','printErr','ALLOC_STACK','ALLOC_STATIC','ALLOC_DYNAMIC','ALLOC_NONE','print']" -s DEMANGLE_SUPPORT=1 -s TOTAL_MEMORY=83886080
    shared:ERROR: invalid forced library: libc++abi

**Comment by kelson42** [comment]
@mossroy @dattaz The upstream ticket has been fixed 4 months ago. Does that mean we could close that ticket?

**Comment by mossroy** [comment]
As you could have read in the latest comments, I did not manage to compile with a recent version of emscripten, so I could not check that this was fixed for us.
It's certainly possible to recompile and test, but I did not have a lot of time to work on that.

In any case, I don't think we should close a ticket if we did not check the problem has been fixed.

It might be something to work on for an upcoming hackathon

**Comment by kelson42** [comment]
@mossroy @Jaifroid Would that be a good time now to check and update this ticket?

**Comment by mossroy** [comment]
I tested again a few weeks ago and did not manage to compile with latest emscripten. So I still can not conclude if this is fixed or not with recent versions of emscripten.
I'll need @mgautierfr help on that, probably at the hackathon.

**Comment by mossroy** [comment]
I just opened https://github.com/emscripten-core/emscripten/issues/15394 with a dedicated test-case
The limit is still at 2 GB for reading bytes.

@mgautierfr : please subscribe to it if they have any questions/remarks on the C code.

**Comment by mossroy** [comment]
I recompiled the test-case with latest emscripten (3.1.12) and ... the problem seems to be solved!
I suppose they somehow fixed it in emscripten, between versions 2.0.32 and 3.1.12

You can try by yourself, as I updated the online version of the test-case:

- https://mossroy.github.io/libzim_wasm/index-bigfile.html is the very simple test-case to check we can read bytes after 2GB
- https://mossroy.github.io/libzim_wasm/index.html to use libzim to access some content (was previously failing with big ZIM files)

**Comment by mossroy** [comment]
I'm quite happy to be able to finally close this issue
